### PR TITLE
feat(CKEditor): add icons and update CSS

### PR DESCRIPTION
**Pull request by langdonholmes**
Minor tweaks to Info CSS and icons for static codeblocks and REPLs.

I tried enabling placeholders for the Info element, but it's creating a fatal error for some reason that escapes me at the moment.

### CODE REVIEW

**Comment by theannachen** [comment]
image
progress pic for the looks so far

**Comment by theannachen** [comment]
@langdonholmes can you go ahead and test it? I know that the codeblock stuff is slightly wacky in some edge cases when entering/indenting or unindenting but it's useable.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Callout/CalloutEditing.js [hunk 1]
```diff
@@ -24,14 +24,14 @@ export default class CalloutEditing extends Plugin {
...
+      allowChildren: "$text",
     });
 
     schema.register("CalloutContent", {
       // Cannot be split or left by the caret.
       isLimit: true,
-      allowIn: ["Callout"],
-      allowChildren: ["paragraph", "listItem"],
+      allowIn: "Callout",
+      allowContentOf: "$root",
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Callout/CalloutEditing.js:34]: What's the purpose of this change?

This breaks the schema by allowing basically anything inside callouts.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  constructor(editor) {
+    super(editor);
+
+    editor.config.define("hljsCodeBlock", {
+      languages: [
+        {
+          language: "plaintext",
+          label: "Plain text",
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:35]: This is not plain text. This is generic code, right? Can you rename this "code" or "no-language" or something?

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+      const res = hljs.highlight(
+        positionParent.getAttribute("language"),
+        textData,
+      );
+      const dom = new DOMParser().parseFromString(
+        "<div>" + res.value + "</div>",
+        "text/xml",
+      );
+
+      //全量更新 重新高亮
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:275]: ????

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockui.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
+/**
+ * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockui.js:4]: Careful with this. If you believe that this code is derivative of someone else's work, you should also make sure that we are in compliance with their license (we are not). GPLv3 _requires_ derivative work to use the same license. We don't have a license yet, and we may or may not use GPLv3.

This looks like boilerplate to me, so I think the license can be removed.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/outdenthljscodeblockcommand.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
+/**
+ * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/outdenthljscodeblockcommand.js:4]: See previous comment.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/utils.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
+/**
+ * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/utils.js:4]: I am concerned that this is now in our commit history.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Warning/WarningEditing.js [hunk 7]
```diff
@@ -24,14 +24,14 @@ export default class WarningEditing extends Plugin {
...
-      allowChildren: "WarningContent",
+      allowChildren: "$text",
     });
 
     schema.register("WarningContent", {
       // Cannot be split or left by the caret.
       isLimit: true,
       allowIn: "Warning",
-      allowChildren: ["paragraph"],
+      allowContentOf: "$root",
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Warning/WarningEditing.js:34]: Why this change? `["paragraph"]` was an intentional choice. Please provide some explanation when you make changes like this!

**Review by langdonholmes — COMMENTED** [review]
Details in the code comments.

Overall, things are looking visually good, but you need to do some housekeeping on the code. Please do not commit licensing statements to our codebase without careful consideration.

Also, some of the changes I made to our editing schemas are broken. Not sure why these changes were made or if it was intentional. Let me know what's going on there please.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+          label: "Python",
+        },
+      ],
+      // A single tab.
+      indentSequence: "\t",
+    });
+  }
+
+  /**
+   * @inheritDoc
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:52]: Please remove these. We don't have any autodocs setup at the moment.

**Comment by langdonholmes** [comment]
Also, can you try to figure out how to get a placeholder working for codeboxes? I think the "Attributes" text that appears above the attributes textbox is ugly and wastes a lot of space.

We can work on making the attributes box collapsible in another PR.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js (hunk 2 above)
Comment (theannachen) [reply src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:35]: I believe that if there's no-language, it will call it undefined.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Callout/CalloutEditing.js (hunk 1 above)
Comment (theannachen) [reply src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Callout/CalloutEditing.js:34]: I think this came from a merge in from main I will revert this

**Comment by theannachen** [comment]
Removed the unnecessary comments and added the placeholder for the static code. Lemme know if there's anything else I should placehold @langdonholmes

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js (hunk 2 above)
Comment (langdonholmes) [reply src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:35]: I think undefined is a better name than "plaintext". My understanding is that this will only be used when the programming language is not specified. If my understanding is correct, then "undefined" is a better description than "plaintext". Right?

It's not super critical, because the programming language will always be defined (currently either JavaScript or Python) when using the CKEditor interface. I'm approving the PR, but please do explain this to me!

**Comment by langdonholmes** [comment]
Anna. This looks great, and I think it's ready to go. I opened it, so you'll need to approve it yourself.

Please take your time with the git merge (or rebase)! Careful not to undo any changes we've merged in since this was opened.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js (hunk 2 above)
Comment (theannachen) [reply src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:35]: The plain text is a separate language in general. I added it because you made a PR with a plain text thing for the static code that can use this. 

If there is not a recognizable language (python, js, plain text) THEN it goes to undefined.

**Comment by theannachen** [comment]
Sounds good! I'll merge in from main first before merging it all to main just in case.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+    schema.extend("$text", {
+      allowIn: "hljsCodeBlock",
+    });
+
+    // Disallow all attributes on $text inside `codeBlock`.
+    schema.addAttributeCheck((context, attributeName) => {
+      if (context.endsWith("hljsCodeBlock $text") && attributeName !== "hljs") {
+        return false;
+      }
+    });
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:104]: This is how the HLJS plugin disables text formatting. You could remove this attribute check and instead add "hljsCodeBlock" to the list of `no_format_elements` in PlainText.js.

This would presumably improve performance by limiting the number of attribute checks that we are running (at the cost of readability).

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js (hunk 2 above)
Comment (langdonholmes) [reply src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:35]: The plaintext plugin is just a utility that disallows text styling on specific elements. It does apply to the "attributes" element of static code blocks.

It's purpose is to prevent users from adding bold/italics/whatever to $text nodes that are children of the following custom elements:
```javascript
    // Prevent text formatting on specific elements
    // We disallow formatting on $text nodes that are direct descendants of these elements.
    const no_format_elements = [
      "InfoTitle",
      "AccordionItemButton",
      "StaticCodeAttributes",
    ];
```

If we do need to use the PlainText plugin for some reason, that would need to be configured by modifying the above list in `PlainText.js`. See my other comment on how the HLJS plugin is already preventing styling in HLJS elements.

Comment (langdonholmes) [reply src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/HLJSCodeBlock/hljscodeblockediting.js:35]: For reference, here is the attribute check that CKEditor runs on the default implementation of codeblocks: https://github.com/ckeditor/ckeditor5/blob/aff1024248ee6fae466c8670cb5b133180277b6c/packages/ckeditor5-code-block/src/codeblockediting.ts#L137

We need some form of attribute check for codeblocks to disable the styling buttons. We already have this, so there's no need to involve the PlainText plugin. Feel free to add some comments to the PlainText plugin or suggest changes that might clarify this point!

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Callout/CalloutEditing.js [hunk 10]
```diff
@@ -24,13 +24,13 @@ export default class CalloutEditing extends Plugin {
       // Behaves like a self-contained block object (e.g. a block image)
       // allowed in places where other blocks are allowed (e.g. directly in the root).
       inheritAllFrom: "$blockObject",
-      allowChildren: ["CalloutContent"],
+      allowChildren: "$text",
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Callout/CalloutEditing.js:27]: This is a regression

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Callout/CalloutEditing.js [hunk 11]
```diff
@@ -24,13 +24,13 @@ export default class CalloutEditing extends Plugin {
...
     });
 
     schema.register("CalloutContent", {
       // Cannot be split or left by the caret.
       isLimit: true,
-      allowIn: ["Callout"],
+      allowIn: "Callout",
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Callout/CalloutEditing.js:33]: Just styling, but also a regression.

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/CodingSandbox/CodingSandboxCommand.js [hunk 12]
```diff
@@ -24,12 +24,12 @@ export default class InsertCodingSandboxCommand extends Command {
 function createCodingSandbox(writer, language) {
   const CodingSandbox = writer.createElement("CodingSandbox");
   const CodingSandboxContent = writer.createElement("CodingSandboxContent");
+  const HLJSCodeBlock = writer.createElement("hljsCodeBlock", {
+    language: `${language.toLowerCase()}`,
+  });
+  console.log(HLJSCodeBlock);
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/CodingSandbox/CodingSandboxCommand.js:30]: leftover console log

File: src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Warning/WarningEditing.js [hunk 13]
```diff
@@ -24,7 +24,7 @@ export default class WarningEditing extends Plugin {
       // Behaves like a self-contained block object (e.g. a block image)
       // allowed in places where other blocks are allowed (e.g. directly in the root).
       inheritAllFrom: "$blockObject",
-      allowChildren: "WarningContent",
+      allowChildren: "$text",
```
Comment (langdonholmes) [inline src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Warning/WarningEditing.js:27]: This is also a regression.

**Review by theannachen — APPROVED** [review]
